app/grandchallenge/algorithms/views.py

```diff
@@ -753,3 +759,50 @@ def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
         context.update({"algorithm": self.base_object})
         return context
+
+
+class AlgorithmAddRepo(
+    LoginRequiredMixin, ObjectPermissionRequiredMixin, UpdateView,
+):
+    model = Algorithm
+    form_class = AlgorithmRepoForm
+    template_name = "algorithms/algorithm_add_repo.html"
+    permission_required = (
+        f"{Algorithm._meta.app_label}.change_{Algorithm._meta.model_name}"
+    )
+    raise_exception = True
+
+    def get_form_kwargs(self):
+        """Return the keyword arguments for instantiating the form."""
+        kwargs = super().get_form_kwargs()
+
+        code = self.request.GET.get("code")
+        repos = []
+        if code is not None:
+            headers = {"Accept": "application/vnd.github.v3+json"}
+
+            resp = requests.post(
+                "https://github.com/login/oauth/access_token",
+                data={
+                    "code": code,
+                    "client_id": settings.GITHUB_CLIENT_ID,
+                    "client_secret": settings.GITHUB_CLIENT_SECRET,
+                },
+                headers=headers,
+            )
+
+            payload = resp.json()
+            headers["Authorization"] = f"token {payload['access_token']}"
+            installations = requests.get(
+                "https://api.github.com/user/installations", headers=headers
+            ).json()
+
+            response = requests.get(
```

Comment:
And here. Are 3 external API calls going to be too much for within a request?

Reply:
It's a bit much. But there's really no way around it, as we cannot do these async. We need the repos before we can return the response. And unfortunately with he way the github API is set up, we need all of these calls to get the repos.